### DISCUSSION TITLE
Comments and some ideas

### DISCUSSION THREAD

**Discussion by Pete-IA** [harbaum/galagino#20] [Ideas]
Hi
I have discovered your project and I find it fascinating. It is what I have been looking for for a long time with the idea of ​​making a fully functional dedicated mini-arcade. I recently tried to port to MAME code from a Teensy to an ESP32, but it was very difficult, perhaps due to lack of knowledge.
After seeing that your code works perfectly on a TTGO T8 1.7 (it's the one I have on hand) I'm designing a 3D printable replica of the Pacman cabinet to start with that (processing).

I'd like to suggest some changes you can make to "improve" the project.

- The TFT's CS pin can be set to GND directly since there is no chip selection.
- Led pin can be directly set to +5V
- The MISO pin also does not need to be connected to the ESP32 board
- If you don't need the led strip, you can connect marquee led directly to +5V

With these changes, which have worked for me, you get some extra pins to add other buttons, like a second action button or start 2 player, put some dip switch...
In the event that more buttons are wanted (which I think is unlikely) several more could be connected to a single (analog) pin using resistors. [Example](https://www.instructables.com/Connecting-Multiple-Buttons-to-a-Single-Pin-on-Ard/)

In another order of things, using a PSRAM, although it is slower, it could run some games even at 15 FPS (more NES emulators run at that frequency and are quite playable. [Example](https://www.instructables.com/Arduino-NES/)

Although I have read your post about how you add games, I have understood the mechanics but I would not know how to implement it. If you could give us some more information I could try it with some other game. Obviously they could not all go together, but they could be made individual or change the set of games.

I have no words to thank you for sharing this project, which I find fascinating.

Sorry for my English because is not my mother language.

**Comment by harbaum** [comment]
You can sure save a few pins here and there. But I had no shortage. E.g. if you wire to LCDs LED pin to constant VCC, then the display is initially on and you see some garbage on startup. If it's off then it's black until it's properly initialized.

You can sure build a two-player version using six more buttons and when doing that your tipps may indeed be helpful.

The NES runs with PSRAM at 15 Hz. But the NES is a single CPU machine while e.g. Galaga runs three CPUs (in fact it runs five, but I emulate only three of these). But yes, it's sure possible to use PSRAM if you e.g. make sure that stack and heap are in internal RAM and only selected data goes into PSRAM.

And finally it probably wouldn't make too much sense to go into more details on how the games were ported. The hardware differs from game to game and no instruction will make this a simple task. You need to understand most aspects of the hardware and you probably also need to be able to debug some Z80 code here and there.  E.g. the multi-CPU-machines (Galaga and Digdug) made me struggle some time with the inter-CPU-communication while others (Dkong and Digdug) took some time to understand the video hardware in all detail.

**Comment by PiArduino** [comment]
First off, great project and fun to build. See WIP perspex build. What I notice whilst testing the games is the hi-score always resetting.
In MAME I remember there's a feature to save the hi-score on exit and load the previous hi-score on boot. Perhaps some use of the ESP32 SPIFF for this in the future?

Again, great work
![20231022_190350](https://github.com/harbaum/galagino/assets/36442329/f321acb0-2b78-46b8-b355-f4f413382f18)
